ingest/src/main/java/com/microsoft/azure/kusto/ingest/ManagedStreamingIngestClient.java

```diff
@@ -13,37 +13,74 @@
 import com.microsoft.azure.kusto.ingest.source.FileSourceInfo;
 import com.microsoft.azure.kusto.ingest.source.ResultSetSourceInfo;
 import com.microsoft.azure.kusto.ingest.source.StreamSourceInfo;
+
+import org.apache.http.client.utils.URIBuilder;
 import org.json.JSONException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.ByteArrayInputStream;
 import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.io.SequenceInputStream;
 import java.lang.invoke.MethodHandles;
 import java.net.URISyntaxException;
+import java.util.UUID;
+
+import sun.misc.IOUtils;
 
 /**
  * <p>ManagedStreamingIngestClient</p>
  * <p>
  * This class combines a managed streaming client with a queued streaming client, to create an optimized experience.
  * Since the streaming client communicates directly with the engine, it's more prone to failure, so this class
  * holds both a streaming client and a queued client.
- * It tries {@value MAX_RETRY_CALLS} times using the streaming client, after which it falls back to the queued streaming client in case of failure.
+ * It tries {@value ATTEMPT_COUNT} times using the streaming client, after which it falls back to the queued streaming client in case of failure.
+ * If the size of the stream is bigger than {@value MAX_STREAMING_SIZE_BYTES}, it will fall back to the queued streaming client.
  * <p>
  * Note that {@code ingestFromBlob} behaves differently from the other methods - since a blob already exists it makes more sense to enqueue it rather than downloading and streaming it, thus ManagedStreamingIngestClient skips the streaming retries and sends it directly to the queued client.
  */
 public class ManagedStreamingIngestClient implements IngestClient {
 
     private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
-    public static final int MAX_RETRY_CALLS = 3;
-    private final QueuedIngestClient queuedIngestClient;
-    private final StreamingIngestClient streamingIngestClient;
+    public static final int ATTEMPT_COUNT = 3;
+    public static final int MAX_STREAMING_SIZE_BYTES = 4 * 1024 * 1024;
+    final QueuedIngestClient queuedIngestClient;
+    final StreamingIngestClient streamingIngestClient;
+    private final ExponentialRetry exponentialRetryTemplate;
+
+    /**
+     * Creates a new ManagedStreamingIngestClient from a DM connection string.
+     * This method infers the engine connection string from the DM connection string.
+     * For advanced usage, use {@link ManagedStreamingIngestClient#ManagedStreamingIngestClient(ConnectionStringBuilder, ConnectionStringBuilder)}
+     * @param dmConnectionString dm connection string
+     * @return a new ManagedStreamingIngestClient
+     */
+    public static ManagedStreamingIngestClient fromDmConnectionString(ConnectionStringBuilder dmConnectionString) throws URISyntaxException {
```

Comment:
I'm sure I'm just missing something, but I don't understand the reason why we'd want a method to infer the other endpoint. In general, we've been moving away from inferring the DM/Engine endpoints from each other, since there are scenarios when it doesn't work. We've been moving towards explicit.

Reply:
We talked about it in the python implementation.
I am very much for this, because for the most common case the user is using the same CSB for both with the only difference is the "ingest-". 
I think it's unnecessary repetition and is prone to bugs (changing in one and missing in the other)

In advanced cases where it's not the user can still explicitly provide both of them, but I think for the common case that's better.

If you're strongly against it we can a have a team discussion (before we release the python version)

Reply:
Nah, I hear this approach. Since we also provide a way to be explicit, I think this approach is good too.